### PR TITLE
Puppet v4 upgrade

### PR DESCRIPTION
**Pull request by bittner**
This PR loosens the dependency on [concat](https://forge.puppet.com/puppetlabs/concat) and [ruby](https://forge.puppet.com/puppetlabs/ruby), and replaces [puppetserver_gem](https://forge.puppet.com/puppetlabs/puppetserver_gem), trying to address the following issues:

- #250 (puppetserver_gem) -- _soft dependency mentioned in [README](https://github.com/DataDog/puppet-datadog-agent#usage)_
- #328 (concat)
- #353 (ruby)

### CODE REVIEW

**Comment by bittner** [comment]
Any plans to merge this PR?

**Comment by truthbk** [comment]
@bittner I would love to merge this as it would clearly help us cleanup.... I'm not 100% sure off-hand but I think this would break for older puppets. We're trying to keep backward compatibility in this `1.x` module. 

But this would certainly be valuable for the next major module release which would be puppet 4+ compatible only.

**Comment by bittner** [comment]
You're not sure. That's one thing.

We're all sure that the current implementation breaks for _newer_ Puppets, though. That is real.

**Comment by truthbk** [comment]
@bittner could you be more specific as to what version this breaks in. I'm not saying we're not going to merge this, I'm saying there's stuff in there that could break for some of our users.

File: metadata.json [hunk 1]
```diff
@@ -70,11 +51,15 @@
...
-      "version_requirement": ">=0.2.0 <1.0.0"
+      "version_requirement": ">=0.2.0 <2.0.0"
     },
     {
       "name": "puppetlabs/concat",
-      "version_requirement": "2.2.0"
+      "version_requirement": ">=2.2.0 <5.0.0"
+    },
+    {
+      "name": "tkishel/system_gem",
```
Comment (truthbk) [inline metadata.json]: This right here is my main concern.

File: metadata.json [hunk 2]
```diff
@@ -11,46 +11,27 @@
...
+      "operatingsystemrelease": ["6", "7", "8", "9"]
     },
     {
       "operatingsystem": "Ubuntu",
-      "operatingsystemrelease": [
-        "10.04",
-        "12.04",
-        "14.04"
-      ]
+      "operatingsystemrelease": ["10.04", "12.04", "14.04", "16.04", "17.04", "17.10"]
```
Comment (truthbk) [inline metadata.json]: We've only officially supported LTS versions thus far - we should probably stay consistent with that (for now)

**Comment by bittner** [comment]
How "more specific" can we be than referencing 3 specific issues that contain the details that motivate these changes?

> that's a no-no for now

When you look at all the other issues and PRs where you say similar things please understand that DataDog users are getting frustrated with the support DataDog is providing for their closed-source platform. We, the "community", are doing free maintenance of this module and all DataDog employees are capable of is saying, _"no, this is not going to be merged"_.

**Find a solution** that is better than ours. **Give suggestions** on how to change the PR so that it also covers all _your_ concerns. Start writing or suggest more tests, **extend the build matrix** to cover the target technologies and versions you care about. -- By saying _"no"_ you're not getting us anywhere, beyond frustration.

**Comment by truthbk** [comment]
@bittner  I did never intend to imply this particular PR was not going to get merged. To the contrary, all I said was that `tkishel/system_gem` was not mergeable due to its implications for older puppets - that was indeed a _suggestion_ on something that could be done to fix the PR as far as we're concerned. The _"no-no"_ was related to that specific element of the PR. The review was meant to be an on-going discussion so we could address the small problems we may have had with it. In fact, that issue was the only big note I had on your PR and was hoping you could address that before we merged it. Your PR is well-thought, well written and will certainly be merged once we address the notes above. 

The test matrix, which is green, covers all puppet versions we're meant to support. The problem is that although the CI does give a sense of when something will definitely break, it will not catch all cases where things _might_ break. Depending on a module that does not support those older puppet versions is something we've been avoiding for a while due to issue we have encountered in the past. The `puppetserver_gem` is indeed one of those cases where a cleaner alternative unfortunately broke the module for our customers.

I apologize for the frustration this may have caused the community and you in particular, we absolutely appreciate the contributions from the community and this module has been driven by you guys from day one.

**Comment by bittner** [comment]
Sorry if I upset you. But to make things clear: Saying somethings is no good is not the same as a suggestion what to better do instead. I replaced the gem-related library for a reason. You should have read the related issue. And I am not alone with this problem.

Please stop making excuses. This only makes things worse. For Datadog. There is clearly passion missing in the maintenance of this module.

**Comment by truthbk** [comment]
@bittner I'm not sure there is a satisfactory solution for the `puppetserver_gem` issue until we drop support for older puppet versions - as far as I can tell any fix will break the module for some portion of our users. You'd be surprised at the number of people still on legacy puppets.

The solution, I believe goes through releasing a new major version of the module that drops support for older puppets. That said, most of the work here would benefit the current module. 

Ideally, in that new module scenario integrations should be resources, as opposed to the current approach. But we could maybe leave that for a yet another major release after that.

**Comment by ardrigh** [comment]
@truthbk 

Can I ask what Datadog's business thinking for supporting unsupported versions of Puppet?
Is there an official time frame for branching to a 2.x release to fix these types of problems?

Version 6 of the Datadog agent is available as a beta, and our company is looking at trying to use it to test (hopefully with this module). 

Now seems like it would be a great time to cut some losses on unsupported versions of Puppet and create branch for release 2.x?  People can submit fixes to a 1.x branch if they need to pin more gems.

The [Puppet Enterprise LTS roadmap](https://puppet.com/misc/puppet-enterprise-lifecycle) lists 2016.4 as the minimum. But this module lists requirement for PE 3.2. And Puppet 2.7 when the current release is 5.3

> Customers on the older inactive versions are strongly encouraged to plan adopting the newer releases to take full benefits of latest features Puppet has to offer. Puppet Enterprise versions 2015.2, 2015.3, 2016.1, 2016.2, 2016.5 and 2017.1 are past the extended support period and will enter the End of Life (EOL) phase as of February 21, 2018. Versions 3.8 and older have been in EOL phase since December 2016. Please contact the technical support team for advice and assistance on moving from an inactive version to a current version of Puppet Enterprise.

I looked at merging our fork with the current release and the Travis failed because another gem requires Ruby 2.1 - so half the test suite automatically failed. That's going to become a losing maintenance game and certainly not something I would bother fixing.

PE 2016.4 runs on Ruby 2.1.9 - that cuts out a good half of the travis.yml lines by itself.

You could also entirely remove the module dependency on `remote_file` by updating to Puppet 4 that supports `http:` for a `file` resource. The currently supported Puppet LTS release supports this. 
Puppet Docs for 4.6 show it is supported for a few versions now.

There would be a few PRs I would submit to fix some problems, but I am really starting to question how useful this module will be if it continues to be pinned to Puppet 2.7

**Comment by bittner** [comment]
@ardrigh Datadog has told me that tomorrow (and every two months) they have community code review at 4pm CET. I'm not sure whether this is a public meeting, though. The meeting location is G+ Hangouts and the #integrations channel of https://datadog.slack.com.

@masci Can you tell us more? Maybe this would be a chance to get good input from the community at large? (Unfortunately, I can't join tomorrow due to a client meeting.)

**Comment by truthbk** [comment]
@ardrigh there are plans in the short-term to drop support for old puppets. What you outline is exactly what we have decided to do. We do have customers on puppet `3.x` and the `1.x` version of the module has been important to maintain.

That said, the plan is to release a new `2.x` version very early in 2018 (targeting late January) to support active puppet versions and drop support for older puppets moving forward. I will be opening a PR very soon.

Although we might backport features if heavily requested (hoping to phase out `1.x` gradually as customers abandon EOL'd puppets), development moving forward will be focused on providing a module relevant in the current context of puppet.

I know it's been a little painful, hopefully we can make them a thing of the past very soon.

Thank you!

**Comment by truthbk** [comment]
@ardrigh on another note, regarding beta 6. There was work merged into this module to support agent6 - docs will get updated ahead of the `1.12.0` release. I have tested the current state of master on puppet5 and it works fine, I presume you might have to cherry-pick some of the work into your fork as opposed to a full merge to get it to fly with your fork if you would rather not use the official module.

Let me know if you need more information or help getting it to work.

**Comment by ardrigh** [comment]
@truthbk thanks for the replies, that's all very helpful information :)

We don't have many changes in our fork at the moment. The previous rpm key changes were integrated into the main release along with the pinned gem fix.

I did see the updates for the beta 6 agent added to this module. Is the 1.12 release due soon? 
We can wait to work off that release.

**Comment by bittner** [comment]
> I did see the updates for the beta 6 agent added to this module

@ardrigh Can you quote the  commit that made this? We're having trouble, it seems, exactly with that addition (see below for the output of an RSpec test we have). That makes it impossible for us to use the current master branch.

```
Evaluation Error: Unknown variable: '::apt_agent6_beta_repo'.
```

**Comment by ardrigh** [comment]
@bittner ah I was just referring to the commit messages for https://github.com/DataDog/puppet-datadog-agent/pull/356

I haven't actually tried to run the current master, I pulled in the tag 1.11 to look at against our changes.
I might have some time to look at the current master later.

**Comment by truthbk** [comment]
#387 has been merged and should address most if not all concerns raised in this PR. It will be released soon (like today). Please open any PRs to address any possible bug or limitation against the latest master after the release. Thank you @bittner 

@ardrigh #387 should allow you to jump on the agent6 bandwagon hopefully without too many issues. Let us know how it goes. As said above, we will be releasing v 2.0 of the module later today.

**Comment by ardrigh** [comment]
@truthbk Thanks for the update. I'll give the new release a spin when it's officially available.

**Pull request closed without merging**